File: dynamic_programming/top_down_longest_common_subsequence.py
import sys


def common_lcs(x, y, idx_of_start, idx_of_end):
    results = [[sys.maxsize for _ in range(len(y) + 1)]
               for _ in range(len(x) + 1)]

    return common_lcs_internal(x, y, idx_of_start, idx_of_end, results)


def common_lcs_internal(x, y, idx_of_start, idx_of_end, results):
    if results[idx_of_start][idx_of_end] < sys.maxsize:
        return results[idx_of_start][idx_of_end]

    result = 0
    if idx_of_start == 0 and idx_of_end == 0:
        result = 1 if x[idx_of_start] == y[idx_of_end] else 0
    elif idx_of_start == 0:
        result = 1 if x[idx_of_start] == y[idx_of_end] \
            else common_lcs_internal(x, y, idx_of_start, idx_of_end - 1, results)
    elif idx_of_end == 0:
        result = 1 if x[idx_of_start] == y[idx_of_end] \
            else common_lcs_internal(x, y, idx_of_start - 1, idx_of_end, results)
    else:
        if x[idx_of_start] == y[idx_of_end]:
            result = common_lcs_internal(x, y, idx_of_start - 1, idx_of_end, results) + 1
        else:
            result = max(common_lcs_internal(x, y, idx_of_start - 1, idx_of_end, results), \
                common_lcs_internal(x, y, idx_of_start, idx_of_end - 1, results))

    results[idx_of_start][idx_of_end] = result

    return result


if __name__ == '__main__':
    x = 'ABCBDAB'
    y = 'BDCABA'

    for i in range(len(x)):
        for j in range(len(y)):
            result = common_lcs(x, y, i, j)

            print(f"i = {i}, j = {j}, X = {x[:i + 1]}, Y = {y[:j + 1]}, result = {result}")

File: dynamic_programming/top_down_longest_common_subsequence.go
//go:build ignore

package main

import (
	"fmt"
	"math"
)

func common_lcs(x, y string, i, j int) int {
	results := make([][]int, max(len(x), len(y)+1))
	for i := 0; i < len(results); i += 1 {
		results[i] = make([]int, max(len(x), len(y)+1))
	}

	for i := 0; i < len(results); i += 1 {
		for j := 0; j < len(results); j += 1 {
			results[i][j] = -math.MaxInt
		}
	}

	return common_lcs_internal(x, y, i, j, results)
}

func common_lcs_internal(x, y string, i, j int, results [][]int) int {
	if results[i][j] > -math.MaxInt {
		return results[i][j]
	}

	result := -math.MaxInt
	if i == 0 && j == 0 {
		if x[i] == y[j] {
			result = 1
		} else {
			result = 0
		}
	} else if i == 0 {
		if x[i] == y[j] {
			result = 1
		} else {
			result = common_lcs_internal(x, y, i, j-1, results)
		}
	} else if j == 0 {
		if x[i] == y[j] {
			result = 1
		} else {
			result = common_lcs_internal(x, y, i-1, j, results)
		}
	} else {
		if x[i] == y[j] {
			result = common_lcs_internal(x, y, i-1, j-1, results) + 1
		} else {
			if common_lcs(x, y, i, j-1) < common_lcs(x, y, i-1, j) {
				result = common_lcs_internal(x, y, i-1, j, results)
			} else {
				result = common_lcs_internal(x, y, i, j-1, results)
			}
		}
	}

	results[i][j] = result

	return result
}

func main() {
	x := "ABCBDAB"
	y := "BDCABA"

	for i := 0; i < len(x); i += 1 {
		fmt.Printf("--- i = %d ---\n", i)
		for j := 0; j < len(y); j += 1 {
			result := common_lcs(x, y, i, j)

			fmt.Printf("(i, j) = (%d, %d) / result = %d\n", i, j, result)
		}
	}
}
